flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriter.java

```diff
@@ -0,0 +1,203 @@
+	}
+
+	private <K> void write(Map<K, List<Long>> offsets, K key, Buffer[] flinkBuffers, boolean precondition) throws Exception {
+		try {
+			if (result.isDone()) {
+				return;
+			}
+			runWithChecks(() -> {
+				checkState(precondition);
+				offsets
+					.computeIfAbsent(key, unused -> new ArrayList<>())
+					.add(checkpointStream.getPos());
+				serializer.writeData(dataStream, flinkBuffers);
+			});
+		} finally {
+			for (Buffer flinkBuffer : flinkBuffers) {
+				flinkBuffer.recycleBuffer();
+			}
+		}
+	}
+
+	void completeInput() throws Exception {
+		LOG.debug("complete input, output completed: {}", allOutputsReceived);
+		complete(!allInputsReceived, () -> allInputsReceived = true);
+	}
+
+	void completeOutput() throws Exception {
+		LOG.debug("complete output, input completed: {}", allInputsReceived);
+		complete(!allOutputsReceived, () -> allOutputsReceived = true);
+	}
+
+	private void complete(boolean precondition, RunnableWithException complete) throws Exception {
+		if (result.isDone()) {
+			// likely after abort - only need to set the flag run onComplete callback
+			doComplete(precondition, complete, onComplete);
+		} else {
+			runWithChecks(() -> doComplete(precondition, complete, onComplete, this::finishWriteAndResult));
+		}
+	}
+
+	private void finishWriteAndResult() throws IOException {
+		dataStream.flush();
+		StreamStateHandle underlying = checkpointStream.closeAndGetHandle();
+		complete(
+				result.inputChannelStateHandles,
+				inputChannelOffsets,
+				(chan, offsets) -> new InputChannelStateHandle(chan, underlying, offsets));
+		complete(
+				result.resultSubpartitionStateHandles,
+				resultSubpartitionOffsets,
+				(chan, offsets) -> new ResultSubpartitionStateHandle(chan, underlying, offsets));
+	}
+
+	private void doComplete(boolean precondition, RunnableWithException complete, RunnableWithException... callbacks) throws Exception {
+		Preconditions.checkArgument(precondition);
+		complete.run();
+		if (allInputsReceived && allOutputsReceived) {
+			for (RunnableWithException callback : callbacks) {
+				callback.run();
+			}
+		}
+	}
+
+	private <I, H extends AbstractChannelStateHandle<I>> void complete(
+			CompletableFuture<Collection<H>> future,
+			Map<I, List<Long>> offsets,
+			BiFunction<I, List<Long>, H> buildHandle) {
+		final Collection<H> handles = new ArrayList<>();
+		for (Map.Entry<I, List<Long>> e : offsets.entrySet()) {
+			handles.add(buildHandle.apply(e.getKey(), e.getValue()));
+		}
+		future.complete(handles);
+		LOG.debug("channel state write completed, checkpointId: {}, handles: {}", checkpointId, handles);
+	}
+
+	private void runWithChecks(RunnableWithException r) throws Exception {
+		try {
+			checkState(!result.isDone(), "result is already completed", result);
+			r.run();
+		} catch (Exception e) {
+			fail(e);
+			throw e;
+		}
+	}
+
+	public void fail(Throwable e) throws Exception {
+		result.fail(e);
+		checkpointStream.close();
+		dataStream.close();
```

Comment:
Are there any problems if `checkpointStream.close()` causes exception to not execute the below `dataStream.close()`?

Reply:
Yes, but in that case closing of `dataStream` doesn't matter.